### PR TITLE
Detect non ha changes | wait for transitions | performance improvements

### PR DESCRIPTION
**Pull request by th3w1zard1**
This is a rewrite designed to accomplish the following:

1. The interval function should not do anything besides update attributes if a transition is in progress.
2. Slightly rework `detect_non_ha_changes` to suit goal 1
3. `_adapt_lights` has been renamed to `_update_manual_control_and_maybe_adapt` and all manual control checks have been moved there.

This pr slightly reworked TurnOnOffListener's `last_state_change` dictionary as it now stores all state changes done to our lights, The old behavior where it is cleared on every new adapt is still there.

Based on #450 

Things this PR fixed that are broken in the main Adaptive Lighting releases ranging from versions 1.3.0 to 1.7.0:
- non HA changes done during a transition might not fire manually controlled events despite `detect_non_ha_changes: true` set. See #447 
- Long transitions are effectively impossible because multiple `light.turn_on` calls with the same data repeatedly shorten the overall transition with each call. See #400
- interval cannot be larger than transitions without `detect_non_ha_changes` breaking.

Things this PR would improve:
- Possibly significant performance increases, especially to those using zigbee/z-wave, as we aren't sending light commands unnecessarily.

### CODE REVIEW

**Comment by th3w1zard1** [comment]
Slow and steady wins the race. Having a local testing environment was substantial for development and cut the testing time down by 90%. Tested on my own system and works great :). @basnijholt this PR is ready for review.

**Comment by th3w1zard1** [comment]
@nathang21 this is ready!

**Comment by th3w1zard1** [comment]
I'm satisfied with this PR, everything is very efficient now. I've been forced to use an interval of 90 or else my zigbee network would go completely out of whack as the update_entity spam was just too much.

Now Adaptive Lighting will correctly wait for all transitions before doing anything, and alt_detect_method finds non ha changes _**without**_ calling update_entity. The interval will now only update attributes, and check if things need to be done.

**Comment by th3w1zard1** [comment]
Performance graphs while using an interval of one second on 5 different lights that are on.

image

Here's performance graphs when using the default 90.

image

File: custom_components/adaptive_lighting/switch.py [hunk 1]
```diff
@@ -665,13 +670,23 @@ def color_difference_redmean(
     return math.sqrt(red_term + green_term + blue_term)
 
 
+def check_direction_change(last: int, current: int, last_adapt_value: int) -> bool:
+    _LOGGER.debug("compare direction: current value %s to last value %s", current, last)
```
Comment (basnijholt) [inline custom_components/adaptive_lighting/switch.py:674]: Wouldn't this break precisely on the top of the parabola? So mid day?

File: custom_components/adaptive_lighting/switch.py [hunk 2]
```diff
@@ -770,6 +821,17 @@ def __init__(
...
+        all_lights = _expand_light_groups(self.hass, self._lights)
+        for light in all_lights:
+            self._manual_lights[light] = {"timer": 0}
+
+        for light in data[CONF_LIGHTS]:
+            if not hasattr(self.turn_on_off_listener, "_switches"):
+                self.turn_on_off_listener._switches = {}
+            self.turn_on_off_listener._switches[
+                light
+            ] = self  # light should never be in multiple switches anyway.
```
Comment (basnijholt) [inline custom_components/adaptive_lighting/switch.py:834]: I have personally have had the same light in multiple switches, and @sizors also reported to be doing that here https://github.com/basnijholt/adaptive-lighting/issues/78#issuecomment-1484502712

File: custom_components/adaptive_lighting/switch.py [hunk 3]
```diff
@@ -1048,14 +1113,18 @@ async def _adapt_light(
...
 
-        if "transition" in features:
-            service_data[ATTR_TRANSITION] = transition
-
         # The switch might be off and not have _settings set.
         self._settings = self._sun_light_settings.get_settings(
             self.sleep_mode_switch.is_on, transition
         )
 
+        # This check is unnecessary.
```
Comment (basnijholt) [inline custom_components/adaptive_lighting/switch.py:1121]: Why is this check unnecessary?

File: tests/test_switch.py [hunk 4]
```diff
@@ -584,6 +584,146 @@ def increased_color_temp():
     assert all([not manual_control[eid] for eid in switch._lights])
 
 
+async def test_non_ha_changes(hass):
```
Comment (basnijholt) [inline tests/test_switch.py:587]: Awesome that you added a test!

File: custom_components/adaptive_lighting/switch.py [hunk 5]
```diff
@@ -1127,6 +1210,74 @@ async def turn_on(service_data):
...
+        last_transition = last_service_data[self._lights[0]][ATTR_TRANSITION]
+        _LOGGER.debug(
+            "%s: wait_transitions: Transition found in last service"
+            "data: %s seconds Timer: %s. Time remaining: %s",
+            self._name,
+            last_transition,
+            self._transition_timer,
+            (last_transition - (perf_counter() - self._transition_timer)),
+        )
+        while (
```
Comment (basnijholt) [inline custom_components/adaptive_lighting/switch.py:1270]: One shouldn't do busy while loops in async code because it will create high CPU work.

Why don't you `await asyncio.sleep` the right amount of time?

**Comment by basnijholt** [comment]
Would it be possible to explain why the current method doesn't work? If so, could you create a test that shows why it fails?

**Comment by th3w1zard1** [comment]
> Would it be possible to explain why the current method doesn't work? If so, could you create a test that shows why it fails?

What current method? Are you referring to the attempt to add a test_non_ha_change?

File: custom_components/adaptive_lighting/switch.py [hunk 6]
```diff
@@ -665,13 +669,23 @@ def color_difference_redmean(
...
+
+
 def _attributes_have_changed(
     light: str,
     old_attributes: dict[str, Any],
     new_attributes: dict[str, Any],
     adapt_brightness: bool,
     adapt_color: bool,
     context: Context,
+    last_adapt_attempt=None,
```
Comment (basnijholt) [inline custom_components/adaptive_lighting/switch.py:688]: It seems like `last_adapt_attempt` is always None in the entire code, so that means we can remove the `if last_adapt_attempt` checks?

File: custom_components/adaptive_lighting/switch.py (hunk 2 above)
Comment (basnijholt) [reply custom_components/adaptive_lighting/switch.py:833]: Just realized that this `turn_on_off_listener._switches` is never used. Please remove.

File: custom_components/adaptive_lighting/switch.py [hunk 7]
```diff
@@ -1078,24 +1148,37 @@ async def _adapt_light(
             color_temp_kelvin = max(min(color_temp_kelvin, max_kelvin), min_kelvin)
             service_data[ATTR_COLOR_TEMP_KELVIN] = color_temp_kelvin
         elif "color" in features and adapt_color:
-            _LOGGER.debug("%s: Setting rgb_color of light %s", self._name, light)
-            service_data[ATTR_RGB_COLOR] = self._settings["rgb_color"]
+            rgb_color = self._settings["rgb_color"]
+            _LOGGER.debug(
+                "%s: Setting rgb_color of light %s to ", self._name, light, rgb_color
+            )
+            service_data[ATTR_RGB_COLOR] = rgb_color
```
Comment (basnijholt) [inline custom_components/adaptive_lighting/switch.py:1155]: This log message only contains 2 `%s` markers. So it is not really logging anything. Since this was likely only used for debugging, can you revert it to the original?

File: tests/test_switch.py [hunk 8]
```diff
@@ -584,6 +584,146 @@ def increased_color_temp():
...
+    # Turn switches on again
+    await switch.adapt_color_switch.async_turn_on()
+    await switch.adapt_brightness_switch.async_turn_on()
+
+    # Check that when no lights are specified, all are reset
+    await change_manual_control(True, {CONF_LIGHTS: switch._lights})
+    assert all([manual_control[eid] for eid in switch._lights])
+    # do not pass "lights" so reset all
+    await change_manual_control(False, {})
+    assert all([not manual_control[eid] for eid in switch._lights])
```
Comment (basnijholt) [inline tests/test_switch.py:724]: I just noticed that this is a 100% carbon copy of the `test_manual_control` test. Why did you add this? This debugging?

File: custom_components/adaptive_lighting/switch.py (hunk 1 above)
Comment (basnijholt) [reply custom_components/adaptive_lighting/switch.py:673]: This function turns out to not be used. So remove :)

**Comment by th3w1zard1** [comment]
@basnijholt Thank you for reviewing this PR! It seems you have a problem with the code involving `alt_detect_method`, would you like me to remove all code regarding that implementation?

**Comment by th3w1zard1** [comment]
> Would it be possible to explain why the current method doesn't work? If so, could you create a test that shows why it fails?

All of my logs are in #447. This pr also fixed the interval > transition problems the old method was having.

**Comment by th3w1zard1** [comment]
Some output of `detect_non_ha_changes` working correctly:
```
2023-03-29 11:48:31.427 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] last_service_data: {'entity_id': 'light.den_ceilingfan_lights', 'brightness': 3, 'color_temp_kelvin': 2702, 'context': <homeassistant.core.Context object at 0x7f6b3c0bef80>}
2023-03-29 11:48:31.428 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] al_den_ceilingfan_lights: 'detect_non_ha_changes: true', calling update_entity(light.den_ceilingfan_lights) and check if it's last adapt succeeded.
2023-03-29 11:48:31.429 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] Brightness of 'light.den_ceilingfan_lights' significantly changed from 3 to 255 with context.id='adapt_lgt:b'h-3=':interval:b'3j''
2023-03-29 11:48:31.429 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] State of 'light.den_ceilingfan_lights' didn't change wrt 'last_service_data' (context.id=adapt_lgt:b'h-3=':interval:b'3j')
2023-03-29 11:48:31.429 DEBUG (MainThread) [custom_components.adaptive_lighting.switch] 'adaptive_lighting.manual_control' event fired for switch.adaptive_lighting_al_den_ceilingfan_lights for light light.den_ceilingfan_lights
```

**Comment by th3w1zard1** [comment]
@basnijholt I fixed `test_significant_change` in this pr.

**Comment by th3w1zard1** [comment]
todo: convert color_temp and rgb_color inside of _attributes_have_changed. Sometimes last_service_data may differ from the actual state if a light is converting color_temp to rgb_color (see #449)

**Comment by th3w1zard1** [comment]
Some pseudo code for kelvin->rgb_color conversions:
```

Set Temperature = Temperature \ 100

Calculate Red:

If Temperature <= 66 Then
	Red = 255
Else
	Red = Temperature - 60
	Red = 329.698727446 * (Red ^ -0.1332047592)
	If Red < 0 Then Red = 0
	If Red > 255 Then Red = 255
End If

Calculate Green:

If Temperature <= 66 Then
	Green = Temperature
	Green = 99.4708025861 * Ln(Green) - 161.1195681661
	If Green < 0 Then Green = 0
	If Green > 255 Then Green = 255
Else
	Green = Temperature - 60
	Green = 288.1221695283 * (Green ^ -0.0755148492)
	If Green < 0 Then Green = 0
	If Green > 255 Then Green = 255
End If

Calculate Blue:

If Temperature >= 66 Then
	Blue = 255
Else

	If Temperature <= 19 Then
		Blue = 0
	Else
		Blue = Temperature - 10
		Blue = 138.5177312231 * Ln(Blue) - 305.0447927307
		If Blue < 0 Then Blue = 0
		If Blue > 255 Then Blue = 255
	End If

End If
```

**Comment by th3w1zard1** [comment]
It looks like someone already made a project for converting colors between each other:

https://github.com/colour-science/colour

**Comment by th3w1zard1** [comment]
I ended up just using the first entry in `last_state_change[light]` as I'd like to test to see if that isn't updated immediately before I do all this unnecessary parsing of `last_service_data[light]`

**Comment by th3w1zard1** [comment]
I ended up doing all the necessary color conversions to fix #449. Added that issue to `test_attributes_have_changed` as well. @basnijholt this is set to auto-merge once/if you approve the changes.

File: custom_components/adaptive_lighting/switch.py [hunk 9]
```diff
@@ -1104,6 +1257,59 @@ async def turn_on(service_data):
...
+        last_transition = last_service_data[found_light][ATTR_TRANSITION]
+        _LOGGER.debug(
+            "%s: wait_transitions: Transition found in last service"
+            "data: %s seconds Timer: %s. Time remaining: %s",
+            self._name,
+            last_transition,
+            self._transition_timer,
+            (last_transition - (perf_counter() - self._transition_timer)),
+        )
+        while (
```
Comment (basnijholt) [inline custom_components/adaptive_lighting/switch.py:1303]: Is the while look still needed now that you are sleeping the right amount of time?

**Comment by th3w1zard1** [comment]
This could be improved by instead starting the transition timer when `TurnOnOffListener` gets the first `state_change` event from our `last_service_data`. This way we'd _ensure_ that the light successfully got our `light.turn_on` command and has started adapting accordingly. I haven't noticed this problem on my Home Assistant setup so I have no way to test it.

**Comment by th3w1zard1** [comment]
@basnijholt Hey, hope you're having a good weekend 😄 

Do you have a development plan for the next release? 
What order are these PRs merging in?
Do you have any issues with this PR in its current form? If not what would you like changed?

I'd like to get the fix '`include_config_in_attributes` for `change_switch_settings`' pushed into a hotfix so that users posting issues could easily use `change_switch_settings` to post their config.

**Comment by basnijholt** [comment]
Hi @th3w1zard1!

> Do you have a development plan for the next release?
> What order are these PRs merging in?
> Do you have any issues with this PR in its current form? If not what would you like changed?


Every single time I look at this PR I get a bit overwhelmed because it changes 20-30% of the lines in `switch.py`. Since it contains 3 changes. Would it be possible to separate them out in to different PRs? Or am I now asking something horrible? 😅

**Comment by th3w1zard1** [comment]
> Hi @th3w1zard1!
> 
> > Do you have a development plan for the next release?
> > What order are these PRs merging in?
> > Do you have any issues with this PR in its current form? If not what would you like changed?
> 
> Every single time I look at this PR I get a bit overwhelmed because it changes 20-30% of the lines in `switch.py`. Since it contains 3 changes. Would it be possible to separate them out in to different PRs? Or am I now asking something horrible? 😅

I absolutely will do that for you, no problem at all! Thanks for the honesty, I should have done this from the start!

**Comment by th3w1zard1** [comment]
@basnijholt Do you mind if I create a new branch on your repo as some of these changes only can merge with this branch.
